Add config-file documentation to RTD (#274)
neuropoly/ivado-medical-imaging: docs/source/configuration_file.md

> # Configuration file

Reviewer: i would definitely reference this file in the `index.rst`, between getting_started.md and tutorials.md.

Author: Done :-)

> * [General parameters](#general-parameters)

Reviewer: by referencing the file in the index.rst, the TOC will be automatically-generated, so it can be removed from here (less maintenance, less prone to errors)

Author: DOne :-)

* [Loader parameters](#loader-parameters)
* [Split dataset](#split-dataset)
* [Training parameters](#training-parameters)
* [Architecture](#architecture)
* [Testing parameters](#testing-parameters)
* [Transformations](#transformations)
* [Examples](#examples)

## General parameters
#### command
Run the specified command. Choices: ``"train"``, ``"test"``, ``"eval"``, to train, test and evaluate a model respectively.
#### gpu
Integer. ID of the GPU to use.
#### log_directory

> Folder name where output files (e.g. trained model, predictions, results) are saved.

Reviewer: ```suggestion
Folder name that will contain the output files (e.g., trained model, predictions, results).
```

#### debugging

> Bool. Allows extended verbosity and intermediate outputs.

Reviewer: ```suggestion
Bool. Extended verbosity and intermediate outputs.
```

## Loader parameters
#### bids_path

> String. Relative path of the BIDS folder.

Reviewer: doesn't it work for absolute paths?

Author: corrected, thanks!

#### target_suffix
Suffix list of the derivative file containing the ground-truth of interest (e.g. [`"_seg-manual"`, `"_lesion-manual"`]). If the list has a length greater than 1, then a multi-class model will be trained.
#### contrasts

> - `train_validation`: list of image modalities loaded for the training and validation. If `multichannel`, then this list represent the different channels of the input tensors. Otherwise, the modalities are mixed and the model has only one input channel.

Reviewer: ```suggestion
- `train_validation`: list of image modalities loaded for the training and validation. If `multichannel=True`, this list represents the different channels of the input tensors. Otherwise, the modalities are mixed and the model only has one input channel.
```

> - `test`: list of image modalities loaded in the testing dataset. Same comment than for `train_validation` regarding `multichannel`.

Reviewer: is it possible to give an example?  people can get confused as to what is expected here. Modality can be interpreted as "image modality", e.g. MRI, ultrasound, CT, etc. I guess you meant "contrast" here? Like "T1w", "T2w", etc.. if so i would give examples

Author: Agree ++. Added suggestion + use contrasts instead of modalities in the code.

> - `balance`: Dict. Used to under-represent some modalities in the dataset: e.g. `{"T1w": 0.1}` will include only 10% of the available `T1w` images into the training/validation/test set. Please set `multichannel` to `false` if you are using this parameter.

Reviewer: ```suggestion
- `balance`: Dict. Enables to weight the importance of specific channels (or contrasts) in the dataset: e.g. `{"T1w": 0.1}` means that only 10% of the available `T1w` images will be included into the training/validation/test set. Please set `multichannel` to `false` if you are using this parameter.
```

#### multichannel

> Bool. Indicated if more than a modality is used by the model. See details in both `train_validation` and `test` for the modalities that are input.

Reviewer: see my comment above about the term "modality". Should we use "channel" everywhere, or "contrast" (maybe the latter is more explicit?)

```suggestion
Bool. Used to specify if more than one channel (or contrast) is used by the model. See details in both `train_validation` and `test` about the input channels.
```

Author: Corrected with contrasts, thanks!

#### slice_axis
Choice between `"sagittal"`, `"coronal"`, and `"axial"`.

> This parameter decides the slice orientation on which the model will train.

Reviewer: ```suggestion
Sets the slice orientation for on which the model will be used.
```

#### slice_filter
Dict:

> - `filter_empty_input`: Bool. Filter empty images if `true`.

Reviewer: ```suggestion
- `filter_empty_input`: Bool. Filter empty images.
```

> - `filter_empty_mask`: Bool. Filter empty GT mask slices if `true`.

Reviewer: i'm a bit confused about two things here:
- what does "filter" mean here? Maybe you can describe
- here the word "slice" is used whereas in the previous line the word "images" is used. Are we talking about the same thing? if so i would use the same term (maybe slices).
```suggestion
- `filter_empty_mask`: Bool. Filter empty ground truth mask slices.
```

Author: Filter: discard a slice from the dataset: ie not used for training etc
`filter_empty_input` discard a slice if all voxel intensities (image not GT) (eg T2w) are all zeros.
I tried to make it clearer. You are right, it was not clear at all.

#### roi
Dict. of parameters about the region of interest

> - `suffix`: suffix of the derivative file containing the ROI used to crop (e.g. `"_seg-manual"`) with `ROICrop` as transform. Please use `null` if you do not want to use a ROI to crop.

Reviewer: not sure i understand how this dict should be defined. An example of the value is given ("_seg-manual"), but what would be the key of the dict?

Author: I tried to clarify it. An example would be:
```json
"roi": {
    "suffix": "_seg-manual",
    "slice_filter_roi": 10
}
```

> - `slice_filter_roi`: int. It filters ROI mask slices with less than this threshold of non zero voxels. Active when using `"ROICrop"` and inputing ROI file.

Reviewer: what is meant by "it filters"?  Did you mean the following?
"Voxels contained in the ROI mask and which value are strictly below the value given in this parameter will be replaced by zero.".

Author: What about:
"It filters (i.e. discards from the dataset) slices where the ROI mask has with less than this number of non-zero voxels."

## Split dataset
#### fname_split

> Filename of a joblib file containing the list of training / validation / test subjects, used to ensure reproducible experiments

Reviewer: ```suggestion
Filename of a joblib file containing the list of training / validation / test subjects. This file can later be used to re-train a model using the same data splitting scheme.
```

#### random_seed
Seed used by the random number generator to split the dataset between training/validation/testing.
#### center_test

> List. List of centers to only include in the testing dataset.

Reviewer: i assume a list of string is expected? we should probably explain where and how the centers are defined (in the participants.tsv file? elsewhere?)

Author: Excellent point. Will correct this.

#### method

> Choice between `"per_patient"` (i.e. shuffle all subjects then splits) or `"per_center"` (split subject according to their acquisition centers).

Reviewer: same comment as before-- where is the definition of a "patient" (is it the suffix of a BIDS-based dataset, e.g. will "sub-03" be considered a "patient") and "center"?

Author: Corrected too. Info extracted from `bids_dataset/participants.tsv`.

#### train_fraction

> Number between `0` and `1` representing the fraction of the dataset used as training set.

Reviewer: mention a float is expected (to follow the convention above, where the parameter type is specified at the beginning of each parameter's description)

#### test_fraction

> Number between `0` and `1` representing the fraction of the dataset used as test set. This parameter is only used if the `method` is `"per_patient"`.

Reviewer: same comment about float

## Training parameters
#### batch_size

> Integer.

Reviewer: ```suggestion
Strictly positive integer.
```

#### loss

> - `name`: Name of the loss function: Choice among the classes that are available [here](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/losses.py).

Reviewer: i would define the hyperlink to the documentation file `losses.md`. I'm not 100% sure how to do the cross-ref in md files, so you'll need to figure out.

Author: Issue opened [here](https://github.com/neuropoly/ivado-medical-imaging/issues/281).

- Other parameters that could be needed in the Loss function definition: see attributes of the Loss function of interest (e.g. `"gamma": 0.5` for `FocalLoss`).
#### training_time

> - `num_epochs`: int.

Reviewer: ```suggestion
- `num_epochs`: Strictly positive integer.
```

> - `early_stopping_epsilon`: Threshold (percentage) for an improvement in the validation loss to be considered meaningful

Reviewer: ```suggestion
- `early_stopping_epsilon`: percentage. Threshold on the validation loss below which training will stop.
```

Author: What about: "If the validation loss difference during one epoch (i.e. `abs(validation_loss[n] - validation_loss[n-1]` where n is the current epoch) is inferior to this epsilon for `early_stopping_patience` consecutive epochs, the training will stop."

> - `early_stopping_patience`: Number of epochs after which the training is stopped if the validation loss improvement not meaningful (i.e. less than `early_stopping_epsilon`)

Reviewer: ```suggestion
- `early_stopping_patience`: Strictly positive integer. Number of epochs after which the training is stopped if the validation loss improvement not meaningful (i.e. less than `early_stopping_epsilon`)
```

#### scheduler
- `initial_lr`: Float. Initial learning rate.
- `scheduler_lr`:
1. `name`: Choice between: `"CosineAnnealingLR"`, `"CosineAnnealingWarmRestarts"` and `"CyclicLR"`. Please find documentation [here](https://pytorch.org/docs/stable/optim.html).
2. Other parameters that are needed for the scheduler of interest (e.g. `"base_lr": 1e-5, "max_lr": 1e-2` for `"CosineAnnealingLR"`).
#### balance_samples

> Bool. If `true`, then positive and negative GT samples are balanced in both training and validation datasets.

Reviewer: ```suggestion
Bool. Balance positive and negative ground truths in both the training and the validation datasets.
```

#### mixup_alpha

> Float. Alpha parameter of the Beta distribution, see [original paper](https://arxiv.org/pdf/1710.09412.pdf).

Reviewer: ```suggestion
Float. Alpha parameter of the Beta distribution, see [original paper](https://arxiv.org/abs/1710.09412).
```

#### transfer_learning

> - `retrain_model`: Filename of the pretrained model (`path/to/pretrained-model`). If `null`, then no transfer learning is performed and the network is trained from scratch.

Reviewer: ```suggestion
- `retrain_model`: Filename of the pretrained model (`path/to/pretrained-model`). If `null`, no transfer learning is performed and the network is trained from scratch.
```

- `retrain_fraction`: Float between 0. and 1. Controls the fraction of the pre-trained model that will be fine-tuned. For instance, if set to 0.5, then the second half of the model will be fine-tuned, while the first layers will be frozen.

## Architecture
Architectures for both segmentation and classification are available and described [here](models.rst).

> If the selected architecture is listed [here](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/loader/loader.py#L14), then a classification task is run, ie the ground-truth are labels extracted from `target`, instead of arrays for the segmentation task.

Reviewer: same comment about redirecting to a file "loader" from the documentation

Author: Yes agree about "L14". So the idea would be to give the link to a RTD page instead of the itself code: Have I understood correctly?

Author: I have opened an issue [here](https://github.com/neuropoly/ivado-medical-imaging/issues/280).

### default_model (Mandatory)
Define the default model (`Unet`) and mandatory parameters that are common to all available architectures (listed [here](models.rst)). If a tailored model is defined (see next section), then the default parameters are merged with the parameters that are specific to the tailored model.
- `name`: `Unet` (default)
- `dropout_rate`: float (e.g. 0.4).
- `batch_norm_momentum`: float (e.g. 0.1).
- `depth`: int, number of down-sampling operations.
Note:
- `in_channel` is automatically defined with `multichannel` and `contrast/training_validation`.
- `out_channel` is automatically defined with `target_suffix`.

> ### Tailored model (optional)

Reviewer: it would clarify to add a space before each new heading

Here are defined the tailored model and the parameters that are specific to it (ie not defined in the default model). See examples:

> - [FiLMedUnet](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/config/config.json#L64)

Reviewer: even when not pointing to another file in the documentation (docs/), i would make a relative link. E.g.: if you create a branch where the json file changes, this absolute link to the master branch will be wrong

Author: Agree! Corrected :-)

Sets the slice orientation for on which the model will be used.

#### slice_filter
Dict. Discard a slice from the dataset if it meets a condition, see below.
- `filter_empty_input`: Bool. Discard slices where all voxel intensities are zeros.
- `filter_empty_mask`: Bool. Discard slices where all voxel labels are zeros.

#### roi
Dict. of parameters about the region of interest
- `suffix`: String. Suffix of the derivative file containing the ROI used to crop (e.g. `"_seg-manual"`) with `ROICrop` as transform. Please use `null` if you do not want to use an ROI to crop.
- `slice_filter_roi`: int. If the ROI mask contains less than `slice_filter_roi` non-zero voxels, the slice will be discarded from the dataset. This feature helps with noisy labels, e.g., if a slice contains only 2-3 labeled voxels, we do not want to use these labels to crop the image. This parameter is only considered when using `"ROICrop"`.

## Split dataset

#### fname_split
String. File name of the log (joblib) that contains the list of training/validation/testing subjects. This file can later be used to re-train a model using the same data splitting scheme.

#### random_seed

> Int. Seed used by the random number generator to split the dataset between training/validation/testing. The use of the same seed ensure the same split between the sub-datasets, which is useful to reproduce results.

Reviewer: ```suggestion
Int. Seed used by the random number generator to split the dataset between training/validation/testing. The use of the same seed ensures the same split between the sub-datasets, which is useful to reproduce results.
```
the use of "Int." and "Integer" is inconsistent, but let's not spend too much additional time on this. At some point, it is possible that the json will be moved to a python class, with docstrings following strict convention for defining types, so no need to be "too" strict at this point

#### balance_samples
Bool. Balance positive and negative labels in both the training and the validation datasets.

#### mixup_alpha
Float. Alpha parameter of the Beta distribution, see [original paper on the Mixup technique](https://arxiv.org/abs/1710.09412).

#### transfer_learning
- `retrain_model`: Filename of the pretrained model (`path/to/pretrained-model`). If `null`, no transfer learning is performed and the network is trained from scratch.
- `retrain_fraction`: Float between 0. and 1. Controls the fraction of the pre-trained model that will be fine-tuned. For instance, if set to 0.5, the second half of the model will be fine-tuned while the first layers will be frozen.

## Architecture
Architectures for both segmentation and classification are available and described in the [Models](models.rst) section.
If the selected architecture is listed in the [loader.py](../../ivadomed/loader/loader.py#L14) file, a classification (not segmentation) task is run. In the case of a classification task, the ground truth will correspond to a single label value extracted from `target`, instead being an array (the latter being used for the segmentation task).

### default_model (Mandatory)

> Dict. Define the default model (`Unet`) and mandatory parameters that are common to all available architectures (listed in the [Models](models.rst) section). If a tailored model is defined (see next section), the default parameters are merged with the parameters that are specific to the tailored model.

Reviewer: "(see next section)" does not work anymore

```suggestion
Dict. Define the default model (`Unet`) and mandatory parameters that are common to all available architectures (listed in the [Models](models.rst) section). For more specific models (see below), the default parameters are merged with the parameters that are specific to the tailored model.
```

# Configuration file

## General parameters

> #### command

Reviewer: i would follow the logical headings order (##, then ###, then ####), otherwise it will impact auto-TOC generation
```suggestion
### command
```

- `test`: list of image contrasts (e.g. `T1w`, `T2w`) loaded in the testing dataset. Same comment than for `train_validation` regarding `multichannel`.
- `balance`: Dict. Enables to weight the importance of specific channels (or contrasts) in the dataset: e.g. `{"T1w": 0.1}` means that only 10% of the available `T1w` images will be included into the training/validation/test set. Please set `multichannel` to `false` if you are using this parameter.

#### multichannel
Bool. Indicated if more than a contrast (e.g. `T1w` and `T2w`) is used by the model. See details in both `train_validation` and `test` for the contrasts that are input.

#### slice_axis
Choice between `"sagittal"`, `"coronal"`, and `"axial"`.
Sets the slice orientation for on which the model will be used.

#### slice_filter
Dict. Discard a slice from the dataset if it meets a condition, see below.
- `filter_empty_input`: Bool. Discard slices where all voxel intensities are zeros.
- `filter_empty_mask`: Bool. Discard slices where all voxel labels are zeros.

#### roi
Dict. of parameters about the region of interest

> - `suffix`: String. Suffix of the derivative file containing the ROI used to crop (e.g. `"_seg-manual"`) with `ROICrop` as transform. Please use `null` if you do not want to use a ROI to crop.

Reviewer: what if a user wants to use an image called "awesomeimage_null.nii"? why not considering an empty string as null?
```suggestion
- `suffix`: String. Suffix of the derivative file containing the ROI used to crop (e.g. `"_seg-manual"`) with `ROICrop` as transform. Please use `null` if you do not want to use an ROI to crop.
```

> - `slice_filter_roi`: int. It filters (i.e. discards from the dataset) slices where the ROI mask has with less than this number of non zero voxels. Active when using `"ROICrop"`.

Reviewer: ```suggestion
- `slice_filter_roi`: int. If the ROI mask contains less than `slice_filter_roi` non-zero voxels, the slice will be discarded from the dataset. This feature helps with noisy labels, e.g., if a slice contains only 2-3 labeled voxels, we do not want to use these labels to crop the image. This parameter is only considered when using `"ROICrop"`.
```

## Split dataset

#### fname_split

> Filename of a joblib file containing the list of training/validation/testing subjects. This file can later be used to re-train a model using the same data splitting scheme.

Reviewer: ```suggestion
String. File name of the log (joblib) that contains the list of training/validation/testing subjects. This file can later be used to re-train a model using the same data splitting scheme.
```

> Filename of a joblib file containing the list of training/validation/testing subjects. This file can later be used to re-train a model using the same data splitting scheme.

Reviewer: is "joblib" a known jargon? does that correspond to a log file?  Is that related to this? https://joblib.readthedocs.io/en/latest/

Author: Not a log file per se: we use it to save numpy, scipy or sklearn objects as alternative to pickle, cf [here](https://scikit-learn.org/stable/modules/model_persistence.html#persistence-example).

Reviewer: ah! so it is related to "https://joblib.readthedocs.io/en/latest/". I would add a hyperlink to "https://joblib.readthedocs.io/en/latest/" to avoid confusion

#### random_seed

> Seed used by the random number generator to split the dataset between training/validation/testing.

Reviewer: what is expected? float? string? any range?

Author: Int is prefered, cf [here](https://docs.python.org/3/library/random.html#random.seed).
I am correcting this.

#### center_test

> List of strings. List of centers to only include in the testing dataset. If used, please include a column `institution_id` in your `bids_dataset/participants.tsv`.

Reviewer: ```suggestion
List of strings. Each string corresponds to an institution/center to only include in the testing dataset (not validation). If used, the file `bids_dataset/participants.tsv` needs to contain a column `institution_id`, which associates a subject with an institution/center.
```

#### method

> Choice between `"per_patient"` (i.e. shuffle all subjects then splits, using the `participant_id` column from `my_bids_dataset/participants.tsv`) or `"per_center"` (split subjects according to their acquisition centers, using the `institution_id` column from `my_bids_dataset/participants.tsv`).

Reviewer: ```suggestion
`{'per_patient', 'per_center'}`. `"per_patient"`: all subjects are shuffled, then split between train/validation/test, regardless their institution. `"per_center"`: all subjects are split so as not to mix institutions between the train, validation and test sets. The latter option enables to ensure the model is working across domains (institutions). Note: the institution information is contained within the `institution_id` column in the `participants.tsv` file.
```

#### train_fraction
Float. Between `0` and `1` representing the fraction of the dataset used as training set.

#### test_fraction

> Float. Between `0` and `1` representing the fraction of the dataset used as test set. This parameter is only used if the `method` is `"per_patient"`.

Reviewer: why couldn't we use that for "per_center"?

Author: When "per_center": the testing dataset is defined by the "center_test", ie "center_test" controls the fraction of the dataset that is used for testing.
While "test_fraction" does this when using "per_patient".
The documentation is not clear at all on that aspect, thank you for raising it. I am suggesting a new wording.

Reviewer: gotcha 😉

## Training parameters

#### batch_size
Strictly positive integer.

#### loss

> - `name`: Name of the loss function: Choice among the classes that are available [here](../../ivadomed/losses.py).

Reviewer: ```suggestion
- `name`: Name of the [loss function Class](../../ivadomed/losses.py).
```

- Other parameters that could be needed in the Loss function definition: see attributes of the Loss function of interest (e.g. `"gamma": 0.5` for `FocalLoss`).

#### training_time
- `num_epochs`: Strictly positive integer.

> - `early_stopping_epsilon`: Float. If the validation loss difference during one epoch (i.e. `abs(validation_loss[n] - validation_loss[n-1]` where n is the current epoch) is inferior to this epsilon for `early_stopping_patience` consecutive epochs, then  the early stopping of the training is triggered.

Reviewer: ```suggestion
- `early_stopping_epsilon`: Float. If the validation loss difference during one epoch (i.e. `abs(validation_loss[n] - validation_loss[n-1]` where n is the current epoch) is inferior to this epsilon for `early_stopping_patience` consecutive epochs, then training stops.
```

> - `early_stopping_patience`: Strictly positive integer. Number of epochs after which the training is stopped if the validation loss improvement not meaningful (i.e. less than `early_stopping_epsilon`)

Reviewer: ```suggestion
- `early_stopping_patience`: Strictly positive integer. Number of epochs after which the training is stopped if the validation loss improvement is smaller than `early_stopping_epsilon`.
```

#### scheduler
- `initial_lr`: Float. Initial learning rate.
- `scheduler_lr`:
1. `name`: Choice between: `"CosineAnnealingLR"`, `"CosineAnnealingWarmRestarts"` and `"CyclicLR"`. Please find documentation [here](https://pytorch.org/docs/stable/optim.html).
2. Other parameters that are needed for the scheduler of interest (e.g. `"base_lr": 1e-5, "max_lr": 1e-2` for `"CosineAnnealingLR"`).

#### balance_samples
Bool. Balance positive and negative labels in both the training and the validation datasets.

#### mixup_alpha

> Float. Alpha parameter of the Beta distribution, see [original paper](https://arxiv.org/abs/1710.09412).

Reviewer: ```suggestion
Float. Alpha parameter of the Beta distribution, see [original paper on the Mixup technique](https://arxiv.org/abs/1710.09412).
```

#### transfer_learning
- `retrain_model`: Filename of the pretrained model (`path/to/pretrained-model`). If `null`, no transfer learning is performed and the network is trained from scratch.

> - `retrain_fraction`: Float between 0. and 1. Controls the fraction of the pre-trained model that will be fine-tuned. For instance, if set to 0.5, then the second half of the model will be fine-tuned, while the first layers will be frozen.

Reviewer: ```suggestion
- `retrain_fraction`: Float between 0. and 1. Controls the fraction of the pre-trained model that will be fine-tuned. For instance, if set to 0.5, the second half of the model will be fine-tuned while the first layers will be frozen.
```

## Architecture

> Architectures for both segmentation and classification are available and described [here](models.rst).
> If the selected architecture is listed [here](../../ivadomed/loader/loader.py#L14), then a classification task is run, ie the ground-truth are labels extracted from `target`, instead of arrays for the segmentation task.

Reviewer: ```suggestion
Architectures for both segmentation and classification are available and described in the [Models](models.rst) section.
If the selected architecture is listed in the [loader.py](../../ivadomed/loader/loader.py#L14) file, a classification (not segmentation) task is run. In the case of a classification task, the ground truth will correspond to a single label value extracted from `target`, instead being an array (the latter being used for the segmentation task).
```

### default_model (Mandatory)

> Define the default model (`Unet`) and mandatory parameters that are common to all available architectures (listed [here](models.rst)). If a tailored model is defined (see next section), then the default parameters are merged with the parameters that are specific to the tailored model.
> - `name`: `Unet` (default)

Reviewer: ```suggestion
Define the default model (`Unet`) and mandatory parameters that are common to all available architectures (listed in the [Models](models.rst) section). If a tailored model is defined (see next section), the default parameters are merged with the parameters that are specific to the tailored model.
- `name`: `Unet` (default)
```

> - `dropout_rate`: float (e.g. 0.4).

Reviewer: ```suggestion
- `dropout_rate`: Float (e.g. 0.4).
```

> - `batch_norm_momentum`: float (e.g. 0.1).

Reviewer: ```suggestion
- `batch_norm_momentum`: Float (e.g. 0.1).
```

> - `depth`: int, number of down-sampling operations.

Reviewer: ```suggestion
- `depth`: Strictly positive integer. Number of down-sampling operations.
```

Note:
- `in_channel` is automatically defined with `multichannel` and `contrast/training_validation`.

> - `out_channel` is automatically defined with `target_suffix`.

Reviewer: how are they automatically defined? what value will/should they have? What do these parameters correspond to?

Author: These parameters do not need to be defined in the config file --> I removed them to avoid confusion. But I explained What and How in `target_suffix` and `contrasts/training_validation`.

> ### Tailored model (optional)

Reviewer: section "default_model" is written with snake_case convention, why not having `tailored_model` here?

Author: `"default_model"` is the actual key of the dict that needs to be present in the config file, eg [here](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/config/config.json#L58).
I agree this is not very clear, I am trying to clarify this.

Reviewer: ahhh! i understand now. OK, what you did (i.e. remove "Tailored model") is now much much clearer.

> Here are defined the tailored model and the parameters that are specific to it (ie not defined in the default model). See examples:

Reviewer: This section includes fancier models and parameters that are not defined in the `default_model`:

Author: I tried to re word this section.

> - [FiLMedUnet](../../ivadomed/config/config.json#L64)

Reviewer: I find it problematic to include a link to a specific line. If the config.json file is updated in the future, i guaranty that someone will forget to update this line number in the documentation, so i would rather remove the link.

>     - `metadata`: choice between `"without"`, `"mri_params"`, and `"contrast"`. If `"mri_params"`, then vectors of [FlipAngle, EchoTime, RepetitionTime, Manufacturer] are input to the FiLM generator. If `"contrast"`, then image contrasts (according to `config/contrast_dct.json`) are input to the FiLM generator.

Reviewer: related note: we use by convention double quote "" for verbose/written text, and single quote '' for variables/dict values (like in this case).
```suggestion
    - `metadata`: `{'without', 'mri_params', 'contrast'}`. `mri_params`: Vectors of [FlipAngle, EchoTime, RepetitionTime, Manufacturer] are input to the FiLM generator. `contrast`: Image contrasts (according to `config/contrast_dct.json`) are input to the FiLM generator.
```

> - [HeMISUnet](../../ivadomed/config/config_spineGeHemis.json#L64)
>     - `missing_contrast`: Bool.

Reviewer: ```suggestion
- `HeMISUnet`
    - `missing_contrast`: Bool.
```

>     - `missing_contrast`: Bool.

Reviewer: any description?

> - [UNet3D](../../ivadomed/config/config_tumorSeg.json#L65)

Reviewer: ```suggestion
- `UNet3D`
```

>     - `length_3D`: tuple indicating the size of the subvolumes or volume used for unet 3D model (depth, width, height).

Reviewer: ```suggestion
    - `length_3D`: (Integer, Integer, Integer). Size of the subvolumes or volume used for unet 3D model: (depth, width, height).
```

>     - `length_3D`: tuple indicating the size of the subvolumes or volume used for unet 3D model (depth, width, height).

Reviewer: what is a "subvolume"?

Author: 3D patch --> I have used this word instead

>     - `length_3D`: tuple indicating the size of the subvolumes or volume used for unet 3D model (depth, width, height).

Reviewer: depth, width and height might be confusing: do they correspond to the first, second and third dimension of a numpy array, respectively? If so i would use that terminology instead (no confusion possible)

Author: What about:
```
- `length_3D`: (Integer, Integer, Integer). Size of the 3D patches used as model's input tensors, along the first, second and third dimension.
```

Reviewer: i think you can even go with: "- `length_3D`: (Integer, Integer, Integer). Size of the 3D patches used as model's input tensors.".  It is fairly self-explanatory that the 1st, 2nd and 3rd arguments refer to the 1st, 2nd and 3rd dimension, although it was not clear before when the terms width, depth, etc. were used (because these terms imply a certain convention on the object itself)

>     - `padding_3D`: size of the overlapping per subvolume and dimensions (e.i `padding:0`). Note: In order to be used, each dimension of an input image needs to be a multiple of length plus 2 * padding and a multiple of 16. To change input image size use the following transformation `CenterCrop3D`. 

Reviewer: i read it 3 times and i still don't understand what this parameter does. It is mentioned "size of the overlapping", but the parameter is called "padding", so is that an overlap (i.e. number of voxels, in each dimension, that overlap between two sliding windows), or a padding (i.e., number of voxels added in each dimension). 

Also, it is still unclear to me what a subvolume is.

"(e.i `padding:0`)" → did you mean "(e.g `padding:0`)"?

to clarify, i would maybe give a concrete example of a volume of size XxYxZ, and what would happen to the subvolume and to the padding applied to the subvolume

Author: Example provided [here](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/loader/loader.py#L423).
This parameter (and its behaviour) is likely to change is a very near future with PR #240 --> I am removing it for now and opening an issue to document it once refactored: issue #279.

>     - `attention_unet`: indicates if attention gates are used in the Unet's decoder.

Reviewer: ```suggestion
    - `attention_unet`: Bool. Use attention gates in the Unet's decoder.
```

## Testing parameters

> - `binarize_prediction`: Indicates if predictions from the model are soft (prediction from 0 to 1) or binarized (either 1 or 0 according to the threshold 0.5).

Reviewer: ```suggestion
- `binarize_prediction`: Bool. Binarize output predictions using a threshold of 0.5. If 'false', output predictions are float between 0 and 1. 
```

#### uncertainty

> Uncertainty computation is performed if `n_it>0` and at least `epistemic` or `aleatoric` is `True`.

Reviewer: I've seen written 'true' and 'True' (same for false/False). I suggest to use a single convention.

Author: Very good point, corrected with `true` and `false` as it is the json convention.

- `epistemic`: Bool.

> - `aleatoric`: Bool.

Reviewer: I would add a short description and a reference for each type. Also, I believe that both cannot be true (or can they)? what happens in this case?

Author: Added short description + refs + clarification than they can both be true (ie both DropOut and Data aug are on).

> - `n_it`: int, number of Monte Carlo iterations.

Reviewer: ```suggestion
- `n_it`: Integer. Number of Monte Carlo iterations. Set to 0 for no uncertainty computation.
```

## Transformations

> Indicate the transformation in the same order you would like them to be applied to your samples. For each transformation, please indicate the parameters as well as the following two optional entries:

Reviewer: ```suggestion
Transformations applied during data augmentation. Transformations are sorted in the order they are applied to the image samples. For each transformation, the following parameters are customizable:
```

- `applied_to`: list betweem `"im", "gt", "roi"`. If not specified, then the transformation is applied to all loaded samples. Otherwise, only applied to the specified types: eg `["gt"]` implies that this transformation is only applied to the ground-truth data.
- `dataset_type`: list between `"training", "validation", "testing"`. If not specified, then the transformation is applied to the three sub-datasets. Otherwise, only applied to the specified subdatasets: eg `["testing"]` implies that this transformation is only applied to the testing sub-dataset.

### Available transformations:
- `NumpyToTensor`
- `CenterCrop2D` (parameters: `size`)
- `ROICrop2D` (parameters: `size`)
- `NormalizeInstance`
- `RandomRotation` (parameters: `degrees`)
- `RandomAffine` (parameters: `translate`)
- `RandomShiftIntensity` (parameters: `shift_range`)
- `ElasticTransform` (parameters: `alpha_range`, `sigma_range`, `p`)
- `Resample` (parameters: `wspace`, `hspace`, `dspace`)
- `AdditionGaussianNoise` (parameters: `mean`, `std`)

> - `DilateGT` (parameters: `dilation_factor`) float, controls the number of iterations of ground-truth dilation depending on the size of each individual lesion, data augmentation of the training set. Use `0` to disable.

Reviewer: if it sets the number of iterations, shouldn't it be integer instead?
```suggestion
- `DilateGT` (parameters: `dilation_factor`) Float. Controls the number of iterations of ground-truth dilation depending on the size of each individual lesion, data augmentation of the training set. Use `0` to disable.
```

- `HistogramClipping` (parameters: `min_percentile`, `max_percentile`)
- `Clage` (parameters: `clip_limit`, `kernel_size`)

> - ` RandomReverse`

Reviewer: ```suggestion
- `RandomReverse`
```

> ## Examples

Reviewer: excellent idea to include this!
